WordPress/Classes/ViewRelated/NUX/LoginViewModel.m

```diff
@@ -0,0 +1,612 @@
+{
+    NSArray *reservedUserNames = @[@"admin",@"administrator",@"root"];
+    return [reservedUserNames containsObject:[[username trim] lowercaseString]];
+}
+
+- (void)toggleSignInFormAction
+{
+    self.shouldDisplayMultifactor = NO;
+    self.userIsDotCom = !self.userIsDotCom;
+    
+    [self.presenter reloadInterfaceWithAnimation:YES];
+}
+
+- (void)displayMultifactorTextField
+{
+    self.shouldDisplayMultifactor = YES;
+    [self.presenter reloadInterfaceWithAnimation:YES];
+    [self.presenter setFocusToMultifactorText];
+}
+
+- (BOOL)isOnePasswordEnabled
+{
+    return [self.onePasswordFacade isOnePasswordEnabled];
+}
+
+- (void)requestOneTimeCode
+{
+    [self.loginFacade requestOneTimeCodeWithLoginFields:[self loginFields]];
+}
+
+- (BOOL)areFieldsValid:(LoginFields *)loginFields
+{
+    if ([self areSelfHostedFieldsFilled:loginFields] && !loginFields.userIsDotCom) {
+        return [self isUrlValid:loginFields.siteUrl];
+    }
+    
+    return [self areDotComFieldsFilled:loginFields];
+}
+
+- (void)setupObservationForAuthenticating
+{
+    [RACObserve(self, authenticating) subscribeNext:^(NSNumber *authenticating) {
+        [self.presenter showActivityIndicator:[authenticating boolValue]];
+    }];
+}
+
+- (void)setupObservationForShouldDisplayMultifactor {
+    [RACObserve(self, shouldDisplayMultifactor) subscribeNext:^(NSNumber *shouldDisplayMultifactor) {
+        BOOL displayMultifactor = [shouldDisplayMultifactor boolValue];
+        if (displayMultifactor) {
+            [self.presenter setUsernameAlpha:LoginViewModelAlphaDisabled];
+            [self.presenter setPasswordAlpha:LoginViewModelAlphaDisabled];
+            [self.presenter setMultiFactorAlpha:LoginViewModelAlphaEnabled];
+        } else {
+            [self.presenter setUsernameAlpha:LoginViewModelAlphaEnabled];
+            [self.presenter setPasswordAlpha:LoginViewModelAlphaEnabled];
+            [self.presenter setMultiFactorAlpha:LoginViewModelAlphaHidden];
+        }
+        
+        self.isUsernameEnabled = !displayMultifactor;
+        self.isPasswordEnabled = !displayMultifactor;
+        self.isMultifactorEnabled = displayMultifactor;
+    }];
+}
+
+- (void)setupObserationForUsernameEnabled
+{
+    [RACObserve(self, isUsernameEnabled) subscribeNext:^(NSNumber *isUsernameEnabled) {
+        [self.presenter setUsernameEnabled:[isUsernameEnabled boolValue]];
+    }];
+}
+
+- (void)setupObservationForPasswordEnabled
+{
+    [RACObserve(self, isPasswordEnabled) subscribeNext:^(NSNumber *isPasswordEnabled) {
+        [self.presenter setPasswordEnabled:[isPasswordEnabled boolValue]];
+    }];
+}
+
+- (void)setupObservationForUserIsDotCom
+{
+    [RACObserve(self, userIsDotCom) subscribeNext:^(NSNumber *userIsDotCom) {
+        BOOL dotComUser = [userIsDotCom boolValue];
+        
+        self.isSiteUrlEnabled = !dotComUser;
+        
+        if (dotComUser) {
+            [self.presenter setPasswordTextReturnKeyType:UIReturnKeyDone];
```

Comment:
Ahhh good point. Moved it here https://github.com/wordpress-mobile/WordPress-iOS/commit/ff14c71e2bc61a920c00eb5b9215fe7745bade70